threeML/plugins/SpectrumLike.py

```diff
@@ -0,0 +1,1495 @@
+
+                    negative_background_n = np.sum(idx)
+
+                    if negative_background_n > 0:
+                        custom_warnings.warn("Generated background has negative counts "
+                                             "in %i channels. Fixing them to zero" % (negative_background_n))
+
+                        randomized_background_counts[idx] = 0
+
+                    # Now compute rates and errors
+
+                    # randomized_background_rate = randomized_background_counts / self.background_exposure
+
+                    randomized_background_count_err = copy.copy(self._back_counts_errors)
+
+                else:
+
+                    raise RuntimeError(
+                        "This is a bug. The combination of source (%s) and background (%s) noise models does not exist" % (
+                        self._observation_noise_model,
+                        self._background_noise_model))
+
+
+
+            # create new source and background spectra
+            # the children of BinnedSpectra must properly override the new_spectrum
+            # member so as to build the appropriate spectrum type. All parameters of the current
+            # spectrum remain the same except for the rate and rate errors
+
+            new_observation = self._observed_spectrum.new_spectrum(new_counts=randomized_source_counts,
+                                                                   new_count_errors=randomized_source_count_err)
+
+
+            new_background = self._background_spectrum.new_spectrum(new_counts=randomized_background_counts,
+                                                                    new_count_errors=randomized_background_count_err)
+
+            # Now create another instance of BinnedSpectrum with the randomized data we just generated
+            # notice that the _new member is a classmethod
+
+            new_spectrum_plugin = self._new(name=new_name,
+                                            observation=new_observation,
+                                            background=new_background,
+                                            verbose=self._verbose,
+                                            **kwargs)
+
+            # Apply the same selections as the current data set
+            if original_rebinner is not None:
+
+                # Apply rebinning, which also applies the mask
+                new_spectrum_plugin._apply_rebinner(original_rebinner)
+
+            else:
+
+                # Only apply the mask
+                new_spectrum_plugin._mask = original_mask
+                new_spectrum_plugin._apply_mask_to_original_vectors()
+
+            # We want to store the simulated parameters so that the user
+            # can recall them later
+
+            new_spectrum_plugin._simulation_storage = clone_model(self._like_model)
+
+            # TODO: nuisance parameters
+
+            return new_spectrum_plugin
+
+    @classmethod
+    def _new(cls,*args,**kwargs):
```

Comment:
They were a way for me to implement a class method inside an instance method. But now I see you type(self)(*args) idea and will switch to that.